akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala

```diff
@@ -693,3 +700,533 @@ private[akka] class BroadcastHub[T](bufferSize: Int) extends GraphStageWithMater
+
+    // Need the queue to be pluggable to be able to use a more performant (less general)
+    // queue in Artery
+    trait PartitionQueue {
+      def init(id: Long): Unit
+      def totalSize: Int
+      def size(id: Long): Int
+      def isEmpty(id: Long): Boolean
+      def nonEmpty(id: Long): Boolean
+      def offer(id: Long, elem: Any): Unit
+      def poll(id: Long): AnyRef
+      def remove(id: Long): Unit
+    }
+
+    object ConsumerQueue {
+      val empty = ConsumerQueue(Queue.empty, 0)
+    }
+
+    final case class ConsumerQueue(queue: Queue[Any], size: Int) {
+      def enqueue(elem: Any): ConsumerQueue =
+        new ConsumerQueue(queue.enqueue(elem), size + 1)
+
+      def isEmpty: Boolean = size == 0
+
+      def head: Any = queue.head
+
+      def tail: ConsumerQueue =
+        new ConsumerQueue(queue.tail, size - 1)
+    }
+
+    class PartitionQueueImpl extends PartitionQueue {
+      private val queues1 = new AtomicReferenceArray[ConsumerQueue](FixedQueues)
+      private val queues2 = new ConcurrentHashMap[Long, ConsumerQueue]
+      private val _totalSize = new AtomicInteger
+
+      override def init(id: Long): Unit = {
+        if (id < FixedQueues)
+          queues1.set(id.toInt, ConsumerQueue.empty)
+        else
+          queues2.put(id, ConsumerQueue.empty)
+      }
+
+      override def totalSize: Int = _totalSize.get
+
+      def size(id: Long): Int = {
+        val queue =
+          if (id < FixedQueues) queues1.get(id.toInt)
+          else queues2.get(id)
+        if (queue eq null)
+          throw new IllegalArgumentException(s"Invalid stream identifier: $id")
+        queue.size
+      }
+
+      override def isEmpty(id: Long): Boolean = {
+        val queue =
+          if (id < FixedQueues) queues1.get(id.toInt)
+          else queues2.get(id)
+        if (queue eq null)
+          throw new IllegalArgumentException(s"Invalid stream identifier: $id")
+        queue.isEmpty
+      }
+
+      override def nonEmpty(id: Long): Boolean = !isEmpty(id)
+
+      override def offer(id: Long, elem: Any): Unit = {
+        @tailrec def offer1(): Unit = {
+          val i = id.toInt
+          val queue = queues1.get(i)
+          if (queue eq null)
+            throw new IllegalArgumentException(s"Invalid stream identifier: $id")
+          if (queues1.compareAndSet(i, queue, queue.enqueue(elem)))
+            _totalSize.incrementAndGet()
+          else
+            offer1() // CAS failed, retry
+        }
+
+        @tailrec def offer2(): Unit = {
+          val queue = queues2.get(id)
+          if (queue eq null)
+            throw new IllegalArgumentException(s"Invalid stream identifier: $id")
+          if (queues2.replace(id, queue, queue.enqueue(elem))) {
+            _totalSize.incrementAndGet()
+          } else
+            offer2() // CAS failed, retry
+        }
+
+        if (id < FixedQueues) offer1() else offer2()
+      }
+
+      override def poll(id: Long): AnyRef = {
```

Comment:
note that there will be at most one concurrent consumer and one concurrent producer for each consumer queue so I don't think the contention is that problematic. ConcurrentHashMap might stripe several to the same lock, but I assume `AtomicReferenceArray` is per element.